File: core/MinimumMovesToEqualArrayElementsII/MinimumMovesToEqualArrayElementsII.py
from typing import List
import random


def minMoves2(nums: List[int]) -> int:
    def quick_select(left, right, target):
        l, r = left, right
        index = random.randint(left, right)
        index_num = nums[index]
        nums[r], nums[index] = nums[index], nums[r]
        index = l
        for i in range(l, r):
            if nums[i] <= index_num:
                nums[index], nums[i] = nums[i], nums[index]
                index += 1
        nums[index], nums[r] = nums[r], nums[index]
        if index == target:
            return nums[index]
        elif index > target:
            return quick_select(left, index - 1, target)
        else:
            return quick_select(index + 1, right, target)

    mid_num = quick_select(0, len(nums) - 1, len(nums) // 2)
    return sum([abs(mid_num - num) for num in nums])



if __name__ == '__main__':
    print(minMoves2([1,10,2,9]))

File: core/MinimumMovesToEqualArrayElementsII/MinimumMovesToEqualArrayElementsII.java
package com.dbc;

import java.util.Random;

public class MinimumMovesToEqualArrayElementsII {
    private void swap(int index1, int index2, int[] nums) {
        int temp = nums[index1];
        nums[index1] = nums[index2];
        nums[index2] = temp;
    }

    private int quickSelect(int left, int right, int target, int[] nums) {
        Random random = new Random();
        int index = random.nextInt(right - left + 1) + left;
        int indexNum = nums[index];
        swap(index, right, nums);
        index = left;
        for (int i = left; i < right; i++) {
            if (nums[i] <= indexNum) swap(i, index++, nums);
        }
        swap(index, right, nums);
        if (index == target) return nums[index];
        else if (index > target) return quickSelect(left, index - 1, target, nums);
        else return quickSelect(index + 1, right, target, nums);
    }

    public int minMoves2(int[] nums) {
        int select = quickSelect(0, nums.length - 1, nums.length / 2, nums);
        int res = 0;
        for (int num : nums) res += Math.abs(num - select);
        return res;
    }
}
